### DISCUSSION TITLE
[Feature] Make immich web app compatible with Samsung TV (tizen OS browser, older Chromium)

### DISCUSSION THREAD

**Discussion by dgtal1** [immich-app/immich#8896] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

immich web app is currently not compatible with Samsung TV (running tizen). I own 2 years old top-class OLED model and unfortunately some things in the web app do not work correctly and some don't work at all.

I filed several bug reports for this:
* https://github.com/immich-app/immich/issues/8893
* https://github.com/immich-app/immich/issues/8883
* https://github.com/immich-app/immich/issues/8881

but later on someone replied that TV browsers are currently not supported.

To me this is a crucial feature to be able to watch photos&videos from my family library on my TV. It's the reason I started seking for a good, self-hosted indexing solution. Photoprism is the closest to fulfill my needs - running well and its UI is TV-compatible, but it fails big time with AI recognition.

I hope the immich team will start supporting TV browsers like the one in Samsung. I think it's a matter of supporting a specific Chromium version - my tizen TV runs Chromium M85 according to these docs - https://developer.samsung.com/smarttv/develop/specifications/web-engine-specifications.html. Also checked its user-agent and indeed it's M85 as stated there.

### Platform

- [ ] Server
- [X] Web
- [ ] Mobile

**Comment by bo0tzz** [comment]
Chromium M85 was released nearly four years ago. I don't think it's feasible to keep supporting browser versions that old.

**Reply by dgtal1** [reply]
In IT 4 years in general is much, I agree. Although I think Chromium and the web standards didn't change that much over 4 years. I understand it's maybe easier to develop using the newest, cutting-edge standards, but the drawback is that only the newest software supports it.
While major PCs and mobile phones browsers (Chrome, Safari, Edge, Firefox) are updated quite frequently on PCs and mobile phones, other devices like TVs, STBs, TV dongles, game consoles are not updated often. At least not in terms of the underlying browser technology  like Chromium. That totally sucks and I wish it wasn't the case, but that's how it looks like in these product categories, which are not mobile phones/tablets nor PCs/laptops. Some software manufacturers do understand it and try to support as much as possible.

In  my opinion making the web app backward-compatible as much as possible (up to a certain and justified level) will make immich compatible with many different platforms and devices. 

I could even start supporting the project if immich only worked on my TV, but without it it's a show stopper for me. I could live without transparent icons and gray UI element backgrounds, but not without a working navigation and watching pics full screen.
I'm hoping some developer with the proper skills will eventually have the same view as me and will implement what's needed to have at least the navigation working.

**Reply by bo0tzz** [reply]
> While major PCs and mobile phones browsers (Chrome, Safari, Edge, Firefox) are updated quite frequently on PCs and mobile phones, other devices like TVs, STBs, TV dongles, game consoles are not updated often.

This is true, but our development first and foremost targets PC and mobile. Yours is the first case I've heard of someone trying to use Immich through a TV browser, for example. Trying to support every device out there is a laudable effort, but it's also a significant amount of work that would directly take away from any other work on Immich.

**Comment by dgtal1** [comment]
> Trying to support every device out there is a laudable effort, but it's also a significant amount of work that would directly take away from any other work on Immich.

I'm not talking about a specific device, but platform - in this case it's Chromium M85 being the basis of tizen's 6.5 web browser. There's a big difference: device vs platform. 

> Yours is the first case I've heard of someone trying to use Immich through a TV browser, for example

But I bet you heard of someone trying to use immich on Android TV, right? (#1614). I'm talking about TV browser, because I know it's not realistic to think of a native app for tizen and I don't think a native app would even add any value. I just think that making something to work on a browser is the least possible effort comparing to a development of a platform-native app from scratch. That's obvious.
Before flutter and other tools of this sort - web apps were the easiest and fastest way to target many platforms - mobile, PCs, anything that's got a capable browser. It was enough to make a web app that's compatible with most browser engines out there and wrap it around a native app with a web view component running the web app. 
I'm not even dreaming about a native app for tizen OS. It's quite niche and cumbersome in terms of development. But it's got quite a modern web engine, that is able to run a lot of things correctly. That's why I think with some (hopefully) minor adjustments immich could run on Samsung TVs' tizen browser correctly (and many things are already working fine, but not the thumb->full screen photo navigation). 

Given people do want to use immich on Android TVs (the thread captured 27 participants which is already a good number) means people want to use immich on TVs in general. And of course people having Android TV won't be uploading photos, but only viewing them or organize them. Same use case as mine.

The facts are that tizen OS despite being a niche itself is the leading TV operating system out there, far ahead of anything else including Android TV. That means there's a huge amount of devices immich could potentially run on additionally to mobile and PC. 

Hopefully adjusting the web app to run on older Chromium is a comparable effort to adjusting the Android app to be compatible with Android TV platform. I'm not able to assess this, but hoping it will happen as it's happening already for Android TV.

**Comment by bo0tzz** [comment]
> That's why I think with some (hopefully) minor adjustments immich could run on Samsung TVs' tizen browser.

That might be the case, but it could also be quite a slog, with this sort of thing it's not necessarily straightforward to assess that up front. It's well possible that the browser build on the TV has some weirdness going on, for example. So to work on this, someone on the Immich team would also need to have (or get) a TV that runs tizen. I don't know of anybody on the team that is interested in doing that right now. Of course PRs are always welcome, it's not like we're opposed to this ;)

> Hopefully adjusting the web app to run on older Chromium is a comparable effort to adjusting the Android app to be compatible with Android TV platform. I'm not able to assess this, but hoping it will happen as it's happening already for Android TV.

The Android TV app is an unofficial effort by a community member, and is not an adjustment of the Immich app but a new codebase altogether.

**Reply by dgtal1** [reply]
>The Android TV app is an unofficial effort by a community member,

And so I'm hoping some day another 'crazy' guy like this one will appear and will PR the support for the tizen browser😊 

I wish I knew enough to do this on my own, but UI technologies are completely not my universe. In the best possible case maybe I would be able to adjust CSS to get non-transparent icons fixed, but it would probably not be in line with the project design rules and I could only do this on my local instance (although I don't even know how could I locally do this only in my container). And it would break as soon as new immich version is published. And as I said - no transparency it not the pain point. Lack of working navigation is, and that's surely beyond my skills to track the reason or fix it, which is a shame as I would even consider making this a local fix and upgrade immich every few months not to reintroduce my local fix every month or so.

**Comment by etasse** [comment]
Here's some extra info about possible workarounds:

Approaches tried:
- [WebOne](https://github.com/atauenis/webone) rewriting proxy: Unable to configure proxy settings for Tizen browser.

Workarounds:
- [BrowService](https://github.com/ttalvitie/browservice): 
Configured the service on local network. Immich is functional but very slow, because web pages are rendered on the service, then re-rendered on the Tizen browser.
Might as well use Tizen's RDP/VNC app to connect to a VM running a recent OS, or use a mobile device's MiraCast to the Smart TV.

_Could WebOne rewriting proxy be integrated into the Immich service and automatically proxy to it based on the User-Agent?_

**Reply by dgtal1** [reply]
Recently my TV updated its platform and now it's based on Chromium 108. Still old, but more compatible with immich. At least videos play and navigation works correctly. But it requires a ton of custom CSS in immich to make it look correctly (Gemini is good enough to produce adjusted CSS working in Chrome 108 based on the Tailwind CSS generated one immich uses. Works for me, but not sure for how long. If Svelte they use for frontend will switch into some cutting edge CSS version then the compatibility will be broken again I guess
